theseus/core/objective.py

```diff
@@ -195,11 +211,18 @@ def add(self, cost_function: CostFunction):
 
         self.cost_functions_for_weights[cost_function.weight].append(cost_function)
 
-        if self.optim_vars.keys() & self.aux_vars.keys():
-            raise ValueError(
-                "Objective does not support a variable being both "
-                "an optimization variable and an auxiliary variable."
-            )
+        for optim_name in self._new_optim_var_name:
```

Comment:
Good catch and thanks for the fix! The overall logic is good. I think it would be better to do this w/o introducing any new member variables. You can compute two sets before line 214, in pseudocode:

optim_vars := cost_function.optim_vars + cost_function.weight.optim_vars
aux_vars := cost_function.aux_vars + cost_function.weight.aux_vars

WIth this you can do the two checks you introduced.